pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1050,15 +1053,23 @@ void messageReceived(MessageIdData messageId, int redeliveryCount, List<Long> ac
 
             // Enqueue the message so that it can be retrieved when application calls receive()
             // if the conf.getReceiverQueueSize() is 0 then discard message if no one is waiting for it.
-            // if asyncReceive is waiting then notify callback without adding to incomingMessages queue
             if (deadLetterPolicy != null && possibleSendToDeadLetterTopicMessages != null &&
                     redeliveryCount >= deadLetterPolicy.getMaxRedeliverCount()) {
                 possibleSendToDeadLetterTopicMessages.put((MessageIdImpl)message.getMessageId(),
                         Collections.singletonList(message));
             }
-            if (peekPendingReceive() != null) {
-                notifyPendingReceivedCallback(message, null);
-            } else if (enqueueMessageAndCheckBatchReceive(message) && hasPendingBatchReceive()) {
+            CompletableFuture<Message<T>> pendingReceiveFuture;
+            lock.writeLock().lock();
+            try {
+                enqueueMessage(message);
```

Comment:
@315157973 
I think we should lock enqueue/notify, dequeue/pending together, otherwise still cause the issue that you mentioned above.

Reply:
@315157973
WDYT?

Reply:
I recommend rolling back directly to avoid unknown problems. Although the performance will drop by 45%.
New changes, CR is hard to see the thread safety problem.
https://github.com/apache/pulsar/pull/9261
Of course, if you can do what @MarvinCai said is lock-free, so much the better.